File: crates/yokan/demo/moods.py
# /// script
# requires-python = ">=3.14"
# ///
"""Enum, Optional and animation. `match` IS pixie's `case` (exhaustiveness checked
natively), the walrus IS `if let some(v)` (Python's own spelling of
narrowing), and `animate=`/`easing=` ride the kernel's animation clock, so
`advance:` frames dump identically in both tiers.
"""
import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), ".."))

from enum import Enum, auto  # noqa: E402

import yokan as ui  # noqa: E402
from yokan import State, store  # noqa: E402


class Mood(Enum):
    HAPPY = auto()
    SAD = auto()


@store
class Tracker:
    last: int | None = None
    trend: Mood = Mood.HAPPY

    def note(self, v: int) -> None:
        self.last = v
        match self.trend:
            case Mood.HAPPY:
                self.trend = Mood.SAD
            case Mood.SAD:
                self.trend = Mood.HAPPY

    def wipe(self) -> None:
        self.last = None


mood: State[Mood] = State(Mood.HAPPY)
sel: State[int | None] = State(None)
note: State[str] = State("-")


def flip():
    match mood():
        case Mood.HAPPY:
            mood.set(Mood.SAD)
        case Mood.SAD:
            mood.set(Mood.HAPPY)


def describe():
    if (v := sel()) is not None:
        note.set(f"picked {v}")
    else:
        note.set("nothing picked")


def view():
    with ui.column(spacing=8, padding=12):
        match mood():
            case Mood.HAPPY:
                ui.text("mood: up", size=18, color="accent", animate=120, easing="out")
            case Mood.SAD:
                ui.text("mood: down", size=18, color="#f38ba8", animate=120, easing="out")
        if (v := sel()) is not None:
            ui.text(f"selection: {v}")
        else:
            ui.text("(no selection)")
        ui.text(f"note: {note()}")
        if (t := Tracker.last) is not None:
            ui.text(f"tracked: {t}", size=12)
        else:
            ui.text("(nothing tracked)", size=12)
        with ui.row(spacing=6):
            ui.button("flip", on_click=flip)
            ui.button("pick", on_click=lambda: sel.set(7))
            ui.button("clear", on_click=lambda: sel.set(None))
            ui.button("describe", on_click=describe)
            ui.button("track", on_click=lambda: Tracker.note(9), animate=100, easing="inOut")
            ui.button("wipe", on_click=Tracker.wipe)


if __name__ == "__main__":
    ui.run(view, title="moods")

File: wakakusa/demo/moods.rb
# Values that are one of a few named things, and a value that may be
# nothing at all. Ruby writes the first as symbols and the second as
# nil, and `case` tells them apart.
require "wakakusa"

class Tracker
  attr_reader :last, :trend

  def initialize
    @last = nil
    @trend = :happy
  end

  def note(v)
    @last = v
    @trend = case @trend
             when :happy then :sad
             else :happy
             end
  end

  def wipe
    @last = nil
  end
end

class Moods
  def initialize
    @mood = :happy
    @sel = nil
    @note = "-"
    @tracker = Tracker.new
  end

  def flip
    @mood = case @mood
            when :happy then :sad
            else :happy
            end
  end

  def describe
    @note = @sel.nil? ? "nothing chosen" : "chose #{@sel}"
  end

  def mood_line
    return text("mood: up", size: 18.0, color: "accent", animate: 120.0, easing: "out") if @mood == :happy

    text("mood: down", size: 18.0, color: "#f38ba8", animate: 120.0, easing: "out")
  end

  def view
    column(spacing: 8.0, padding: 12.0) {
      mood_line
      if @sel.nil?
        text "(no selection)"
      else
        text "selection: #{@sel}"
      end
      text "note: #{@note}"
      if @tracker.last.nil?
        text "(nothing tracked)", size: 12.0
      else
        text "tracked: #{@tracker.last}", size: 12.0
      end
      row(spacing: 6.0) {
        button("flip") { flip }
        button("pick") { @sel = 7 }
        button("clear") { @sel = nil }
        button("describe") { describe }
        button("track", animate: 100.0, easing: "inOut") { @tracker.note(9) }
        button("wipe") { @tracker.wipe }
      }
    }
  end
end

run(Moods.new, title: "moods")
